sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+              withResource(builtTable.filter(trackerTab.getColumn(0))) { filterTab =>
+                GpuColumnVector.from(filterTab, GpuColumnVector.extractTypes(batch))
+              }
+            }
+          }
+        }
+        // Combine right columns with null columns for left side
+        val ret = withResource(filteredBatch) { rightBatch =>
+          val numFilterRows = rightBatch.numRows()
+          if (numFilterRows > 0) {
+            val leftColumns = streamAttributes.map { attr =>
+              GpuColumnVector.fromNull(numFilterRows, attr.dataType)
+            }
+            withResource(new ColumnarBatch(leftColumns.toArray, numFilterRows)) { leftBatch =>
+              Some(GpuColumnVector.combineColumns(leftBatch, rightBatch))
+            }
+          } else {
+            None
+          }
+        }
+        rightTracker = None
+        ret
+      }
+    }
+  }
+
+  private def trueColumnTable(numRows: Int): Table = {
+    withResource(Scalar.fromBool(true)) { trueScalar =>
+      withResource(ai.rapids.cudf.ColumnVector.fromScalar(trueScalar, numRows)) {
+        new Table(_)
+      }
+    }
+  }
+
+  // Create a boolean column to indicate which gather map rows are valid.
+  private def validIndexMask(gatherView: ColumnView): ColumnView = {
+    withResource(Scalar.fromInt(-numRightRows)) { minValue =>
+      withResource(Scalar.fromInt(numRightRows)) { maxValue =>
+        withResource(gatherView.greaterOrEqualTo(minValue)) { greaterEqualMin =>
+          withResource(gatherView.lessThan(maxValue)) { lessMax =>
+            lessMax.and(greaterEqualMin)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Update the tracking mask.
+   */
+  private def updateTrackingMask(rightGatherMap: LazySpillableGatherMap): Unit = {
+    // Get the current tracking table, or all true table to start with
+    val rightTrackingTable = rightTracker.map { spillableBatch =>
+      withResource(spillableBatch.releaseBatch()) { trackingBatch =>
+        GpuColumnVector.from(trackingBatch)
+      }
+    }.getOrElse {
+      trueColumnTable(numRightRows)
+    }
+    // Filter the right gather map to remove invalid indices
+    val numGatherMapRows = rightGatherMap.getRowCount.toInt
+    val filteredGatherMap = {
+      withResource(rightGatherMap.toColumnView(0, numGatherMapRows)) { gatherView =>
+        withResource(gatherView.copyToColumnVector()) { gatherVec =>
+          withResource(new Table(gatherVec)) { gatherTab =>
+            withResource(validIndexMask(gatherView)) { mask =>
+              gatherTab.filter(mask)
+            }
+          }
+        }
+      }
+    }
+    // Update all hits in the gather map with false (no longer needed) in the tracking table
+    val updatedTrackingTable = {
+      withResource(rightTrackingTable) { trackingTable =>
+        withResource(Scalar.fromBool(false)) { falseScalar =>
+          withResource(filteredGatherMap) { filteredMap =>
```

Comment:
filteredGatherMap can leak if `fromBool` just above throws

Reply:
I will fix this.

Reply:
Fixed.